contents/tutorials/prevent-fouc-ab-tests.md

```diff
@@ -0,0 +1,74 @@
+---
+title: How to prevent Flash of Unstyled Content (FOUC) during A/B tests
+date: 2024-11-08
+author:
+  - daniel-bachhuber
+showTitle: true
+sidebar: Docs
+tags:
+  - experimentation
+---
+
+It's cruel enough to subject your visitors to an A/B test, but it's even worse if they know about it!
+
+The "Flash of Unstyled Content" (FOUC) happens when users briefly see the control variant before the test variant loads. It's a common issue when running experiments on website building platforms, and happens when the control variant briefly renders before the test variant loads.
+
+If you're unable to [bootstrap the feature flags into the document](/docs/feature-flags/bootstrapping), which generally requires access to server-side code, the best thing you can do is briefly delay the page from displaying until the experiment code executes.
+
+Here's a short example as to how you might do this.
+
+## 1. Prevent the page from displaying
+
+First, you'll need to prevent the page from displaying while the PostHog JavaScript library loads.
+
+Before the `</head>` tag, add the following CSS definition:
+
+```html
+<style>
+.ph-hide-element {
+    opacity: 0 !important;
+}
+</style>
+```
+
+This code snippet defines a new CSS rule that applies 0 opacity to any element with a `ph-hide-element` class. You might want to add it after the PostHog web snippet so you remember why it's there.
```

Comment:
"You might want to add it after the PostHog web snippet so you remember why it's there."
I dont understand what this means or why its helpful?

Reply:
> I dont understand what this means

It's a suggestion to put the `<style>` tag after the PostHog web snippet:

```html
<script>
    !function(t,e){var o,n,p,r;e.__SV||(window.posthog=e,e._i=[],e.init=function(i,s,a){function g(t,e){var o=e.split(".");2==o.length&&(t=t[o[0]],e=o[1]),t[e]=function(){t.push([e].concat(Array.prototype.slice.call(arguments,0)))}}(p=t.createElement("script")).type="text/javascript",p.crossOrigin="anonymous",p.async=!0,p.src=s.api_host.replace(".i.posthog.com","-assets.i.posthog.com")+"/static/array.js",(r=t.getElementsByTagName("script")[0]).parentNode.insertBefore(p,r);var u=e;for(void 0!==a?u=e[a]=[]:a="posthog",u.people=u.people||[],u.toString=function(t){var e="posthog";return"posthog"!==a&&(e+="."+a),t||(e+=" (stub)"),e},u.people.toString=function(){return u.toString(1)+".people (stub)"},o="init capture register register_once register_for_session unregister unregister_for_session getFeatureFlag getFeatureFlagPayload isFeatureEnabled reloadFeatureFlags updateEarlyAccessFeatureEnrollment getEarlyAccessFeatures on onFeatureFlags onSessionId getSurveys getActiveMatchingSurveys renderSurvey canRenderSurvey getNextSurveyStep identify setPersonProperties group resetGroups setPersonPropertiesForFlags resetPersonPropertiesForFlags setGroupPropertiesForFlags resetGroupPropertiesForFlags reset get_distinct_id getGroups get_session_id get_session_replay_url alias set_config startSessionRecording stopSessionRecording sessionRecordingStarted captureException loadToolbar get_property getSessionProperty createPersonProfile opt_in_capturing opt_out_capturing has_opted_in_capturing has_opted_out_capturing clear_opt_in_out_capturing debug".split(" "),n=0;n<o.length;n++)g(u,o[n]);e._i.push([i,s,a])},e.__SV=1)}(document,window.posthog||[]);
    posthog.init('<ph_project_api_key>', {api_host: 'https://us.i.posthog.com', person_profiles: 'identified_only'})
</script>
<style>
.ph-hide-element {
    opacity: 0 !important;
}
</style>
```

> why its helpful?

Marketing sites often end up with a bunch of junk in the `<head>`, and it can be difficult to figure out why it's there.

I went with inline comments instead: 1fc454441